File: finance_app/storage/seed.py
from dataclasses import dataclass
from datetime import date

from finance_app.domain.constants import (
    ENTRY_FIXED,
    ENTRY_INSTALLMENT,
)


@dataclass(frozen=True)
class SeedEntry:
    type: str
    name: str
    amount_cents: int
    category: str
    start_date: date
    installments: int = 1
    notes: str = "Migrado do Notion"


INITIAL_ENTRIES = (
    SeedEntry(ENTRY_FIXED, "Vivo", 4_300, "Assinatura", date(2026, 2, 1)),
    SeedEntry(ENTRY_FIXED, "Mei", 8_090, "Outros", date(2026, 2, 1)),
    SeedEntry(ENTRY_FIXED, "YouTube Premium", 2_690, "Assinatura", date(2026, 2, 1)),
    SeedEntry(ENTRY_FIXED, "Crunchyroll", 1_999, "Assinatura", date(2026, 2, 1)),
    SeedEntry(ENTRY_FIXED, "Google One", 999, "Assinatura", date(2026, 2, 1)),
    SeedEntry(ENTRY_FIXED, "Cap cut", 3_290, "Assinatura", date(2026, 2, 1)),
    SeedEntry(ENTRY_INSTALLMENT, "Celular", 323_100, "Outros", date(2026, 1, 1), 9),
    SeedEntry(ENTRY_INSTALLMENT, "Cadeira gamer", 77_924, "Casa", date(2026, 1, 1), 4),
    SeedEntry(ENTRY_INSTALLMENT, "Jogo Pokemon", 25_732, "Entretenimento", date(2026, 1, 1), 4),
    SeedEntry(ENTRY_INSTALLMENT, "Viagem", 121_704, "Entretenimento", date(2026, 1, 1), 2),
    SeedEntry(ENTRY_INSTALLMENT, "Silent hill F", 22_038, "Entretenimento", date(2026, 1, 1), 2),
    SeedEntry(ENTRY_INSTALLMENT, "Resident evil", 35_180, "Entretenimento", date(2026, 4, 1), 4),
)

File: netlify/functions/lib/seed.ts
import { eq } from "drizzle-orm";
import { db } from "../../../db/index.js";
import { settings, categories, users } from "../../../db/schema.js";
import {
  ACTIVE_YEAR,
  CATEGORIES,
  ENTRY_FIXED,
  ENTRY_INSTALLMENT,
  INITIAL_INVESTED_CENTS,
} from "./constants.js";
import { createEntry } from "./finance.js";

export const DEFAULT_USERNAME = "gamoxkun";
export const DEFAULT_PASSWORD_HASH = "611fd55829d3540c2bc56b6c27777c8f19bfb63aee2ff595ac0b9160ed200ad3";

const INITIAL_ENTRIES = [
  { type: ENTRY_FIXED, name: "Vivo", amountCents: 4300, category: "Assinatura", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_FIXED, name: "Mei", amountCents: 8090, category: "Outros", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_FIXED, name: "YouTube Premium", amountCents: 2690, category: "Assinatura", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_FIXED, name: "Crunchyroll", amountCents: 1999, category: "Assinatura", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_FIXED, name: "Google One", amountCents: 999, category: "Assinatura", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_FIXED, name: "Cap cut", amountCents: 3290, category: "Assinatura", startDate: "2026-02-01", installments: 1 },
  { type: ENTRY_INSTALLMENT, name: "Celular", amountCents: 323100, category: "Outros", startDate: "2026-01-01", installments: 9 },
  { type: ENTRY_INSTALLMENT, name: "Cadeira gamer", amountCents: 77924, category: "Casa", startDate: "2026-01-01", installments: 4 },
  { type: ENTRY_INSTALLMENT, name: "Jogo Pokemon", amountCents: 25732, category: "Entretenimento", startDate: "2026-01-01", installments: 4 },
  { type: ENTRY_INSTALLMENT, name: "Viagem", amountCents: 121704, category: "Entretenimento", startDate: "2026-01-01", installments: 2 },
  { type: ENTRY_INSTALLMENT, name: "Silent hill F", amountCents: 22038, category: "Entretenimento", startDate: "2026-01-01", installments: 2 },
  { type: ENTRY_INSTALLMENT, name: "Resident evil", amountCents: 35180, category: "Entretenimento", startDate: "2026-04-01", installments: 4 },
] as const;

export async function initializeDefaultsForUser(
  userId: number,
  initialInvestedCents = 0
): Promise<void> {
  for (const name of CATEGORIES) {
    await db
      .insert(categories)
      .values({ userId, name })
      .onConflictDoNothing();
  }
  await db
    .insert(settings)
    .values({ userId, key: "active_year", value: String(ACTIVE_YEAR) })
    .onConflictDoNothing();
  await db
    .insert(settings)
    .values({
      userId,
      key: "initial_invested_cents",
      value: String(initialInvestedCents),
    })
    .onConflictDoNothing();
}

export async function seedDefaultUserIfNeeded(): Promise<void> {
  await db
    .insert(users)
    .values({ username: DEFAULT_USERNAME, passwordHash: DEFAULT_PASSWORD_HASH })
    .onConflictDoNothing();

  const [defaultUser] = await db
    .select({ id: users.id })
    .from(users)
    .where(eq(users.username, DEFAULT_USERNAME))
    .limit(1);

  if (!defaultUser) {
    throw new Error("Usuario padrao nao encontrado.");
  }

  await initializeDefaultsForUser(defaultUser.id, INITIAL_INVESTED_CENTS);

  // Try to claim the seed lock atomically
  const result = await db
    .insert(settings)
    .values({ userId: defaultUser.id, key: "seeded_initial_data", value: "1" })
    .onConflictDoNothing()
    .returning();

  if (result.length === 0) return; // Already seeded by a prior invocation

  for (const entry of INITIAL_ENTRIES) {
    await createEntry(
      defaultUser.id,
      entry.type,
      entry.name,
      entry.amountCents,
      entry.category,
      entry.startDate,
      entry.installments
    );
  }
}
